### PR TITLE
Make asset and log file uploads non-blocking

### PR DESCRIPTION
**Pull request by kraih**
Over the past few days i've been looking into making all the uploader parts of `OpenQA::Worker::Job` non-blocking. While a rewrite of the entire uploader will probably be necessary in the future for performance improvements, it is also possible to perform uploads in a subprocess for now. That has the advantage that uploads no longer block the worker event loop, allowing it to keep sending status updates to the webui.

I also tried to make image uploads non-blocking in the same way, but unfortunately the test `24-worker-job.t` depends very heavily on timing bugs that make it impossible to change without removing the whole test (i've attempted to fix it, but there are just too many bugs).

This can also be seen as a proof of concept for moving asset/log file uploading into a Minion job, since the subprocess already behaves very similarly.

### CODE REVIEW

**Comment by kraih** [comment]
I will be deploying this branch to the staging machines for acceptance testing.

File: t/lib/OpenQA/Test/Utils.pm [hunk 1]
```diff
@@ -381,4 +379,11 @@ sub client_output {
     return $out;
 }
 
+sub shared_hash {
+    my $hash = shift;
+    state $file = do { my $f = tempfile; lock_store {}, $f->to_string; $f };
+    return lock_retrieve $file->to_string unless $hash;
+    lock_store $hash, $file->to_string;
+}
```
Comment (kraih) [inline t/lib/OpenQA/Test/Utils.pm]: I don't love this, but it was the most minimalistic solution i could think of for read/write sharing a data structure between two processes.

**Comment by kraih** [comment]
I've also changed some callbacks not to be optional, That's preparation for the future, callbacks should never be optional when you're writing continuation-passing style code.

**Comment by okurz** [comment]
Looks ok so far. I know we talked about this in chat but I am not exactly sure what the actual change in behaviour is. I have read the PR description and the commit messages but neither explain what actual problem we have this fixes. Is it related to https://progress.opensuse.org/issues/55238 that the worker is happily uploading but the web UI considers the host dead because of missing status updates? Because there the problem looks more like related to the test result files and thumbnails and not the "additional" log files. I don't know if another process would help there. Am I reading this wrong?

**Comment by kraih** [comment]
@okurz Yes, that is related. It only helps with the log file case though. I could not fix thumbnail uploading yet for the reasons mentioned above. Performing uploads with a separate process means that the main worker process can perform other tasks in the meantime, such as sending status updates to the webui.

You have to understand though that the worker is fundamentally broken by design. It is a huge pile of hacks that have been added one by one over many years. Each adding a new layer of bugs. And we will have to redesign it from the ground up to have any chance of making it work reliably. This PR is only the beginning and there will probably be many more followups that won't make sense to you right away.

**Comment by okurz** [comment]
I trust you have the great master plan in mind :) However when merging we have to consider the impact on our existing instances. It might be a compromise for the time being, sure. Could you also try to increase the code coverage?

Regarding tests on a staging instance: I recommend you bring up the load of the web UI system to at least 30 and see how it behaves, e.g. if the web UI host is not fast enough to process uploads by the workers, etc. This could be done either by doing something heavy on the web UI host, e.g. write continuously from /dev/zero or /dev/urandom to local files (without completely depleting the free space) or add so many workers which have nothing much to actually test but upload a lot of files, e.g. 10k little log files, and see how it behaves?

**Comment by kraih** [comment]
I'm also working on another branch now that makes all uploads non-blocking. And it seems to work pretty well so far, just the test coverage is lower since it's incompatible with `24-worker-job.t`. https://github.com/os-autoinst/openQA/compare/master...kraih:async_thumbnails

**Comment by kraih** [comment]
Closing this PR, since #2270 is replacing it.

**Pull request closed without merging**